### PR TITLE
54 add widget type field and break the current widget into smaller components

### PR DESCRIPTION
**Pull request by longshuicy**
Post definition with supported field widget name will now automatically populate the frontend to construct such form.
Supported widgetType includes `Select`, `TextField`, `DateTimePicker`. If not supported widgetType was provided, the frontend fallback is `TextField`

 e.g.
```
{
    "name" : "Unit",
    "description" : "Unit",
    "context" : {
        "doi" : "https://schema.org/QuantitativeValue"
    },
    "fields" : [
        {
            "name" : "unit",
            "list" : false,
            "widgetType": "Select",
            "config": {
                 "type" : "enum",
                 "options": ["Ampere", "Kelvin", "Second"]
            },
            "required" : true
        }
    ]
}
```
More examples see `backend/data/metadata/definitions` folder

image

### CODE REVIEW

**Comment by longshuicy** [comment]
Ready to be reviewed

**Comment by tcnichol** [comment]
Question, I was testing this with a file trying to add metadata of each type. For coordinates, I notice that if I try to input this:
Screen Shot 2022-08-23 at 3 42 31 PM

it ends up like this:

Screen Shot 2022-08-23 at 3 42 37 PM

am i using that metadat field wrong?

**Comment by longshuicy** [comment]
> Question, I was testing this with a file trying to add metadata of each type. For coordinates, I notice that if I try to input this: Screen Shot 2022-08-23 at 3 42 31 PM
> 
> it ends up like this:
> 
> Screen Shot 2022-08-23 at 3 42 37 PM
> 
> am i using that metadata field wrong? I made sure that I had the most recent one from /data/metadata in my db.

might be related to this specific field is a list of tuples; the backend might not parse it correctly... let me investigate a bit

**Comment by longshuicy** [comment]
> > Question, I was testing this with a file trying to add metadata of each type. For coordinates, I notice that if I try to input this: Screen Shot 2022-08-23 at 3 42 31 PM
> > it ends up like this:
> > Screen Shot 2022-08-23 at 3 42 37 PM
> > am i using that metadata field wrong? I made sure that I had the most recent one from /data/metadata in my db.
> 
> might be related to this specific field is a list of tuples; the backend might not parse it correctly... let me investigate a bit

I realize this need to have a dynamic input widgets on the frontend; might not worth the effort for now since it's just a demonstration.

File: backend/data/metadata/definitions/coordinates.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+            "name" : "point",
+            "list" : true,
+            "widgetType": "Map",
+            "config": {
+                "type" : "tuple"
+            },
+            "required" : true
+        }
+    ]
+}
```
Comment (longshuicy) [inline backend/data/metadata/definitions/coordinates.json:18]: Let's not test this for now. This needs to work with an interactive map widget

**Review by max-zilla — APPROVED** [review]
Read through the changes and tested on a fresh database by adding the definitions in data/metadata (except coordinates) and going through each one in a new dataset adding+editing, it all looked good.

**Comment by max-zilla** [comment]
looks like there's a merge conflict on .gitignore

**Pull request merged**